gslib/tests/test_rsync.py

```diff
@@ -3130,3 +3131,52 @@ def test_kms_key_specified_will_not_prevent_non_kms_copy_to_s3(self):
     # formatting (i.e. specifying KMS key in a request to S3's API).
     with SetBotoConfigForTest([('GSUtil', 'prefer_api', 'json')]):
       self.RunGsUtil(['rsync', tmp_dir, suri(bucket_uri)])
+
+  def test_bucket_to_bucket_includes_arbitrary_headers(self):
+    bucket1_uri = self.CreateBucket()
+    bucket2_uri = self.CreateBucket()
+    self.CreateObject(bucket_uri=bucket1_uri,
+                      object_name='obj1',
+                      contents=b'obj1')
+
+    stderr = self.RunGsUtil([
+        '-D', '-h', 'arbitrary:header', 'rsync',
+        suri(bucket1_uri),
+        suri(bucket2_uri)
+    ],
+                            return_stderr=True)
+
+    headers_for_all_requests = re.findall(r"Headers: \{([\s\S]*?)\}", stderr)
+    self.assertTrue(headers_for_all_requests)
+    for headers in headers_for_all_requests:
+      self.assertIn("'arbitrary': 'header'", headers)
+
+  @SkipForGS('Tests that S3 SSE-C is handled.')
+  def test_s3_sse_is_handled_with_arbitrary_headers(self):
+    tmp_dir = self.CreateTempDir()
+    tmp_file = self.CreateTempFile(tmpdir=tmp_dir, contents=b'foo')
+    bucket_uri1 = self.CreateBucket()
+    bucket_uri2 = self.CreateBucket()
+
+    header_flags = [
+        '-h',
+        '"x-amz-server-side-encryption-customer-algorithm:AES256"',
+        '-h',
+        '"x-amz-server-side-encryption-customer-key:MTIzNDU2Nzg5MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTI="',
+        '-h',
+        '"x-amz-server-side-encryption-customer-key-md5:dnF5x6K/8ZZRzpfSlMMM+w=="',
+    ]
+
+    with SetBotoConfigForTest([('GSUtil', 'check_hashes', 'never')]):
+      self.RunGsUtil(header_flags +
+                     ['-m', 'cp', tmp_file,
```

Comment:
Does this actually ensure more than one thread is used?

Reply:
Not necessary for the test, removed.

Reply:
Then how are you testing headers get to worker threads?

Reply:
Rsync runs in parallel by default (that's why the customer was running into issues with that command and not others): https://github.com/GoogleCloudPlatform/gsutil/blob/master/gslib/commands/rsync.py#L1004

Reply:
Do we not have more direct tests for the `Command`, `WorkerPool`, and `WorkerThread`?

Reply:
Not really, none that are set up to catch cases like this, or need to be updated after the API change (other than the ones that I have changed)